modules/core/test/image/testImageWarpAffine.cpp

```diff
@@ -0,0 +1,288 @@
+/****************************************************************************
+ *
+ * This file is part of the ViSP software.
+ * Copyright (C) 2005 - 2017 by Inria. All rights reserved.
+ *
+ * This software is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * ("GPL") version 2 as published by the Free Software Foundation.
+ * See the file LICENSE.txt at the root directory of this source
+ * distribution for additional information about the GNU GPL.
+ *
+ * For using ViSP with software that can not be combined with the GNU
+ * GPL, please contact Inria about acquiring a ViSP Professional
+ * Edition License.
+ *
+ * See http://visp.inria.fr for more information.
+ *
+ * This software was developed at:
+ * Inria Rennes - Bretagne Atlantique
+ * Campus Universitaire de Beaulieu
+ * 35042 Rennes Cedex
+ * France
+ *
+ * If you have questions regarding the use of this file, please contact
+ * Inria at visp@inria.fr
+ *
+ * This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
+ * WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * Description:
+ * Test for vpImageTools::warpAffine() function.
+ *
+ * Author:
+ * Vikas Thamizharasan
+ *
+ *****************************************************************************/
+/*!
+  \example testImageWarpAffine.cpp
+
+  \brief Test the warpAffine function with different types of transformation matrix
+*/
+
+#include <visp3/gui/vpDisplayD3D.h>
+#include <visp3/gui/vpDisplayGDI.h>
+#include <visp3/gui/vpDisplayGTK.h>
+#include <visp3/gui/vpDisplayOpenCV.h>
+#include <visp3/gui/vpDisplayX.h>
+#include <visp3/io/vpImageIo.h>
+#include <visp3/io/vpParseArgv.h>
+
+#include <visp3/core/vpIoTools.h>
+#include <visp3/core/vpImageTools.h>
+
+// List of allowed command line options
+#define GETOPTARGS "cdip:h"
+
+
+namespace {
+  /*
+    Print the program options.
+
+    \param name : Program name.
+    \param badparam : Bad parameter name.
+    \param ipath: Input image path.
+    \param ppath: Personal image path.
+   */
+  void usage(const char *name, const char *badparam, std::string ipath, std::string ppath)
+  {
+    fprintf(stdout, "\n\
+  Test image warping.\n\
+  \n\
+  SYNOPSIS\n\
+    %s [-i <input image path>] [-p <personal image path>] [-c] [-d]\n\
+       [-h]\n                 \
+  ", name);
+
```

Comment:
We should consider command line options with more than one character. This is possible using `vpParseArgv` class

`GETOPTARGS` and usage() should be removed.
Then `getOption()` should be modified like:

```
  /*!
    Set the program options.

    \param argc : Command line number of parameters.
    \param argv : Array of command line parameters.
    \param opt_ipath: Input image path.
    \param opt_ppath: Personal image path.
    \param opt_display : Do not display if set.
    \param opt_click : Do not need click if set.
    \param opt_angle : .............
    \param opt_scale_x : .............
    \param opt_scale_y : .............
    \param opt_translation_x : .............
    \param opt_translation_y : .............
    \return false if the program has to be stopped, true otherwise.
  */
  bool getOptions(int argc, const char **argv, std::string &opt_ipath, std::string &opt_ppath,
                  bool &opt_display, bool &opt_click, float &opt_angle,
                  float &opt_scale_x, float &opt_scale_y,
                  float &opt_translation_x, float &opt_translation_y)
  {
    char *opt_ipath_ = NULL;
    char *opt_ppath_ = NULL;

    vpParseArgv::vpArgvInfo argTable[] =
    {
      {"-i", vpParseArgv::ARGV_STRING, (char*) NULL, (char *) &opt_ipath_,
       "Set image input path. From this path read \"ViSP-images/Klimt/Klimt.pgm\""},
      {"-p", vpParseArgv::ARGV_STRING, (char*) NULL, (char *) &opt_ppath_,
       "Path to an image used to test image warping."},
      {"-c", vpParseArgv::ARGV_CONSTANT, 0, (char *) &opt_click,
       "Disable mouse click."},
      {"-d", vpParseArgv::ARGV_CONSTANT, 0, (char *) &opt_display,
       "Disable display."},
      {"-angle", vpParseArgv::ARGV_FLOAT, (char*) NULL, (char *) &opt_angle,
       "Angle in degrees."},
      {"-scale_x", vpParseArgv::ARGV_FLOAT, (char*) NULL, (char *) &opt_scale_x,
       "Scale factor along x."},
      {"-scale_y", vpParseArgv::ARGV_FLOAT, (char*) NULL, (char *) &opt_scale_y,
       "Scale factor along y."},
      {"-trans_x", vpParseArgv::ARGV_FLOAT, (char*) NULL, (char *) &opt_translation_x,
       "Translation along x."},
      {"-trans_y", vpParseArgv::ARGV_FLOAT, (char*) NULL, (char *) &opt_translation_y,
       "Translation along y."},
      {"-h", vpParseArgv::ARGV_HELP, (char*) NULL, (char *) NULL,
       "Print the help."},
      {(char*) NULL, vpParseArgv::ARGV_END, (char*) NULL, (char*) NULL, (char*) NULL}
    };
    // Read the command line options
    if(vpParseArgv::parse(&argc, argv, argTable,
                          vpParseArgv::ARGV_NO_LEFTOVERS |
                          vpParseArgv::ARGV_NO_ABBREV |
                          vpParseArgv::ARGV_NO_DEFAULTS)) {
      return (false);
    }

    if (opt_ipath_)
      opt_ipath = std::string(opt_ipath_);
    if (opt_ppath_)
      opt_ppath = std::string(opt_ppath_);

    return true;
  }
```

The `main()` becomes:

```
    std::string env_ipath;
    std::string opt_ipath;
    std::string ipath;
    std::string opt_ppath;
    std::string ppath;
    std::string filename;
    bool opt_display = true;
    bool opt_click = true;
    float opt_angle = 0;
    float opt_scale_x = 1;
    float opt_scale_y = 1;
    float opt_translation_x = 0;
    float opt_translation_y = 0;

....

    // Read the command line options
    if (getOptions(argc, argv, opt_ipath, opt_ppath, opt_display, opt_click,
                   opt_angle, opt_scale_x, opt_scale_y, opt_translation_x, opt_translation_y) == false) {
      exit (EXIT_FAILURE);
    }

....

    // Make Tansformation Matrix
    T[0][0] =  opt_scale_x * cos(vpMath::rad(opt_angle)); T[0][1] = opt_scale_x * sin(vpMath::rad(opt_angle));  T[0][2] = 0;
    T[1][0] = -opt_scale_y * sin(vpMath::rad(opt_angle)); T[1][1] = opt_scale_y * cos(vpMath::rad(opt_angle));  T[1][2] = 0;
    T[2][0] =  opt_translation_x;                         T[2][1] = opt_translation_y;                          T[2][2] = 1;


```